examples/ct/structures.go

```diff
@@ -99,15 +112,208 @@ func SignV1SCTForPrecertificate(km crypto.KeyManager, cert *x509.Certificate, t
+	leaf := ct.MerkleTreeLeaf{Version: ct.V1, LeafType: ct.TimestampedEntryLeafType, TimestampedEntry: timestampedEntry}
 
 	return serializeAndSignSCT(km, leaf, sctInput, t)
 }
 
-func getSCTForSignatureInput(t time.Time) (ct.SignedCertificateTimestamp) {
+func getSCTForSignatureInput(t time.Time) ct.SignedCertificateTimestamp {
 	return ct.SignedCertificateTimestamp{
-		SCTVersion:ct.V1,
+		SCTVersion: ct.V1,
 		Timestamp:  uint64(t.UnixNano() / millisPerNano), // spec uses millisecond timestamps
 		Extensions: ct.CTExtensions{}}
-}
\ No newline at end of file
+}
+
+func NewLogEntry(leaf ct.MerkleTreeLeaf, certChain []*x509.Certificate) *CTLogEntry {
+	chain := []ct.ASN1Cert{}
+
+	for _, cert := range certChain {
+		chain = append(chain, cert.Raw)
+	}
+
+	logEntry := new(CTLogEntry)
+
+	logEntry.Leaf = leaf
+	logEntry.Chain = chain
+
+	return logEntry
+}
+
+// WriteTimestampedEntry writes out a TimestampedEntry structure in the binary format defined
+// by RFC 6962. The CT go code includes a deserializer but not a serializer so we might as
+// well make this available.
+func WriteTimestampedEntry(w io.Writer, t ct.TimestampedEntry) error {
+	if err := binary.Write(w, binary.BigEndian, &t.Timestamp); err != nil {
+		return err
+	}
+	if err := binary.Write(w, binary.BigEndian, &t.EntryType); err != nil {
+		return err
+	}
+	switch t.EntryType {
+	case ct.X509LogEntryType:
+		if err := writeVarBytes(w, t.X509Entry, ct.CertificateLengthBytes); err != nil {
+			return err
+		}
+	case ct.PrecertLogEntryType:
+		if err := binary.Write(w, binary.BigEndian, t.PrecertEntry.IssuerKeyHash); err != nil {
+			return err
+		}
+		if err := writeVarBytes(w, t.PrecertEntry.TBSCertificate, ct.PreCertificateLengthBytes); err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("unknown EntryType: %d", t.EntryType)
+	}
+
+	return writeVarBytes(w, t.Extensions, ct.ExtensionsLengthBytes)
+}
+
+// WriteMerkleTreeLeaf writes a MerkleTreeLeaf in the binary format specified by RFC 6962.
+// The CT go code includes a deserializer but not a serializer and we might as well make this
+// available to other users.
+func WriteMerkleTreeLeaf(w io.Writer, l ct.MerkleTreeLeaf) error {
+	if l.Version != ct.V1 {
+		return fmt.Errorf("unknown Version: %d", l.Version)
+	}
+
+	if l.LeafType != ct.TimestampedEntryLeafType {
+		return fmt.Errorf("unknown LeafType: %d", l.LeafType)
+	}
+
+	if err := binary.Write(w, binary.BigEndian, l.Version); err != nil {
+		return err
+	}
+	if err := binary.Write(w, binary.BigEndian, l.LeafType); err != nil {
+		return err
+	}
+	if err := WriteTimestampedEntry(w, l.TimestampedEntry); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Serialize writes out a CTLogEntry in binary form. This is not an RFC 6962 data structure
+// and is only used internally by the log.
+func (c CTLogEntry) Serialize(w io.Writer) error {
```

Comment:
I think I'm becoming more convinced that I should put the Write() stuff back into CT.